irohad/main/server_runner.cpp

```diff
@@ -39,36 +47,68 @@ iroha::expected::Result<int, std::string> ServerRunner::run() {
   builder.SetMaxReceiveMessageSize(INT_MAX);
   builder.SetMaxSendMessageSize(INT_MAX);
 
-  serverInstance_ = builder.BuildAndStart();
-  serverInstanceCV_.notify_one();
+  server_instance_ = builder.BuildAndStart();
+  server_instance_cv_.notify_one();
 
   if (selected_port == 0) {
     return iroha::expected::makeError(
-        (boost::format(kPortBindError) % serverAddress_).str());
+        (boost::format(kPortBindError) % server_address_).str());
   }
 
   return iroha::expected::makeValue(selected_port);
 }
 
 void ServerRunner::waitForServersReady() {
-  std::unique_lock<std::mutex> lock(waitForServer_);
-  while (not serverInstance_) {
-    serverInstanceCV_.wait(lock);
+  std::unique_lock<std::mutex> lock(wait_for_server_);
+  while (not server_instance_) {
+    server_instance_cv_.wait(lock);
+  }
+}
+
+iroha::expected::Result<std::shared_ptr<grpc::ServerCredentials>, std::string>
+ServerRunner::createSecureCredentials() {
+  grpc::SslServerCredentialsOptions::PemKeyCertPair keypair = {
+      tls_keypair_->private_key, tls_keypair_->certificate};
+  auto options = grpc::SslServerCredentialsOptions();
+  options.pem_key_cert_pairs.push_back(keypair);
+  return iroha::expected::makeValue(grpc::SslServerCredentials(options));
+}
+
+boost::optional<std::string>
+ServerRunner::addListeningPortToBuilder(grpc::ServerBuilder &builder, int *selected_port) {
+  if (tls_keypair_) {  // if specified, requested to enable TLS
+    std::shared_ptr<grpc::ServerCredentials> credentials;
+    boost::optional<std::string> credentials_error = {};
+    createSecureCredentials().match(
+        [&](const auto &secure_credentials) {
+          credentials = secure_credentials.value;
+        },
+        [&](const auto &error) { credentials_error = error.error; });
+    if (credentials_error) {
+      return credentials_error;
+    }
+
+    builder.AddListeningPort(server_address_, credentials, selected_port);
```

Comment:
How are the keys having incorrect format handled? I am afraid, an exception may arise on this line or somewhere later. If this is the case, should catch such an exception and provide the user with an appropriate error message.
Also please check what happens if we provide a syntactically correct key that does not match the certificate.

Reply:
I could not yet find how to handle this error getting a cause from gRPC, but invalid formats and mismatched certificates would both give `selected_port == 0`, and would print an error message, such as `E0714 18:22:43.053279000 4790195648 ssl_transport_security.cc:675]     Invalid cert chain file.` What would you suggest, do I keep digging (I have literally spent the weekend looking for a solution), or this would be sufficient?

Reply:
I think that if you checked various error cases and irohad 1) points up the correct error reason and 2) terminates gracefully - that is enough for now. Thank you for devoting your time to the investigation!